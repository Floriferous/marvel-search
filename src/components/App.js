import React from 'react';
import { Provider } from 'react-redux';

import '../css/app.css';
import configureStore from '../store';
import HomePage from './HomePage';

const store = configureStore();

const App = () => (
  <Provider store={store}>
    <HomePage />
  </Provider>
);

export default App;
