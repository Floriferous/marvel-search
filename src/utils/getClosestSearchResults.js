import { createSearchKey } from '../actions/search';

/**
 * getClosestSearchResults - Searches for the closest search results
 * currently in memory. It iterates on a search string, and removes one
 * character at a time until it finds an old search result to display.
 *
 * If the first run passes, it simply returns the exact search results.
 *
 * An isLoading flag is also returned, to indicate whether old or exact
 * results are currently being returned
 *
 * @param {Object} searchResults Object with all previous search results
 * @param {String} search        current search string
 * @param {String} pagination    current pagination
 *
 * @return {Object}
 */
const getClosestSearchResults = (searchResults, search, pagination) => {
  let broaderResults;
  let isExactSearch;

  // Return immediately if search length equals 1 or pagination is larger than 0
  if (search.length === 1 || pagination > 0) {
    return {
      data: searchResults[createSearchKey(search, pagination)] || {},
      isLoading: false, // taken care of by CharacterList
    };
  }

  for (let i = 0; i < search.length; i += 1) {
    const broaderSearch = search.slice(0, search.length - i);
    broaderResults = searchResults[createSearchKey(broaderSearch, pagination)];
    if (broaderResults) {
      isExactSearch = i === 0;
      break;
    }
  }
  return { data: broaderResults || {}, isLoading: !isExactSearch };
};

export default getClosestSearchResults;
