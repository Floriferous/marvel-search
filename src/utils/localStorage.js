export const storageAvailable = () => {
  try {
    const storage = localStorage;
    const x = '__storage_test__';
    storage.setItem(x, x);
    storage.removeItem(x);
    return true;
  } catch (e) {
    return false;
  }
};

export const saveValueAtKey = (key, value) => {
  if (storageAvailable()) {
    localStorage.setItem(key, JSON.stringify(value));
    return true;
  }

  return false;
};

export const getValueAtKey = (key) => {
  if (storageAvailable()) {
    return JSON.parse(localStorage.getItem(key));
  }

  return false;
};
