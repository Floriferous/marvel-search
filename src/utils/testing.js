import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

export const testComponent = (Component, props) => {
  if (!testComponent.mountedLockScreen) {
    testComponent.mountedLockScreen = Enzyme.shallow(<Component {...props} />);
  }
  return testComponent.mountedLockScreen;
};

testComponent.reset = () => {
  testComponent.mountedLockScreen = undefined;
};

export const { shallow, mount, render } = Enzyme;
export default Enzyme;
