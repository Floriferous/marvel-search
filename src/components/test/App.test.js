import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { testComponent } from '../../utils/testing';
import App from '../App';

describe('App', () => {
  const component = () => testComponent(App);

  beforeEach(() => {
    testComponent.reset();
  });

  it('matches its snapshot', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
  });

  it('renders a Provider', () => {
    expect(component().find('Provider').length).toBe(1);
  });

  describe('the Provider', () => {
    let Provider;
    beforeEach(() => {
      Provider = component().find('Provider');
    });

    it('is given 2 props', () => {
      expect(Object.keys(Provider.props()).length).toBe(2);
    });

    it('is given a store prop', () => {
      expect(Provider.props().store).toBeDefined();
    });

    it('is given HomePage as a child', () => {
      expect(Provider.props().children).toBeDefined();
    });
  });
});
