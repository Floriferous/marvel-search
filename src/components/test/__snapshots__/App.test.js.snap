// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches its snapshot 1`] = `
<main>
  <h1>
    Marvel Search
  </h1>
  <input
    autoFocus={true}
    onChange={[Function]}
    placeholder="Spider-man"
    type="text"
    value=""
  />
  <h2
    className="empty"
  >
    Start searching or add bookmarks to find your heroes
  </h2>
  <div
    className="disclaimer"
  >
    <p>
      Data provided by Marvel. Â© 2014 Marvel
    </p>
  </div>
</main>
`;
